Update Coproducts

abergfeld opened the pull request:
Fixes #1284 

- Deprecates `cop` and `coproductOf` methods as they always need generic types declared due to ambiguity 
- Add `first`, `second`, `third` etc constructor methods that don't require generic types declared unless they cannot be inferred
- Make the data classes public as an alternative constructor 
- Update documentation to include runnable code examples
- Update documentation to include example of flattening hierarchies
- Update code generation to include class level and method level documentation

abergfeld commented:
@JorgeCastilloPrz @raulraja Just some questions:

- Should we use the generic letter for extension method names `"String".a()` or the name of the data class `"String".first()`?

- Should `coproductOf` also be deprecated? It should suffer from the safe inference issue

- Should I add something to the docs about the data classes being a valid way to construct a Coproduct?

- Is there something more specific we should use for the deprecation message?

raulraja commented:
I'd make the extensions the same name as the data class constructors. 
`corpoducOf` should also be deprecated as you noted. 
It'd be nice if the docs showcased both construction styles. ADT constructors and syntax.
The docs should be moved to kdocs and relinked in the menu to the generated api doc, ideally with runnable snippets as it is done in other places like functor, Bracket, etc.

abergfeld commented:
@raulraja Updated to use the `first`, `second` etc naming. Also deprecated the `coproductOf` functions as well in favor of just using the data classes. 

I added the ADT examples to the docs and added kdocs for all the non-deprecated things. 

I have a couple things I'm not sure of:
1) Given that there's so many functions related to Coproducts (2-22 sealed classes, etc) I just added runnable snippets to the current docs page. I assume it's not worth linking the same kind of runnable snippets over and over if it's generated off the kdocs. Just wanted to validate that it's fine to leave that out of the kdocs.

2) I'm having some trouble building the documentation locally. The code snippets all show up correctly but when I run them, it says that they don't compile on things that didn't break the build. Does the code snippet running happen with the current master version of Arrow? I had compilation failures running Ank locally and fixed them but the snippets still failed after I ran the jeykll command

JorgeCastilloPrz commented:
Are you following the steps described [here](https://github.com/arrow-kt/arrow/blob/master/modules/docs/arrow-docs/README.md#how-to-deploy-docs-to-a-local-server)?
 (dokka + ank + deploy)

JorgeCastilloPrz commented on modules/core/arrow-generic/src/test/kotlin/arrow/generic/CoproductTest.kt:
```diff
@@ -82,5 +82,37 @@ class CoproductTest : UnitSpec() {
+            val secondValue = 100L
+            val thirdValue = none<String>()
+
+            firstValue.cop<String, Long, Option<String>>() shouldBe firstValue.first<String, Long, Option<String>>()
```
didn't we remove cop ext funs?

abergfeld replied:
No I think the intent is to deprecate them and remove them in a later version? https://github.com/arrow-kt/arrow/issues/1284#issuecomment-460337722

I'm not sure what the next Arrow version will be I guess, I may have misunderstood

raulraja replied:
that is correct @abergfeld they will be deprecated in 0.9.0 coming out now and will be removed sometime before 1.0 final.

JorgeCastilloPrz commented on modules/core/arrow-generic/src/test/kotlin/arrow/generic/CoproductTest.kt:
```diff
@@ -82,5 +82,37 @@ class CoproductTest : UnitSpec() {
+        }
+
+        "types can be inferred with the cop replacement functions" {
+            fun typeInferenceTest(input: String?): Coproduct3<String, Int, Option<String>> {
```
would we need to add a test for inference when direcly using the ADT constructors in the function returns?

JorgeCastilloPrz commented on modules/core/arrow-generic/src/test/kotlin/arrow/generic/CoproductTest.kt:
```diff
@@ -82,5 +82,37 @@ class CoproductTest : UnitSpec() {
+            val secondValue = 100L
+            val thirdValue = none<String>()
+
+            First<String, Long, Option<String>>(firstValue) shouldBe firstValue.first<String, Long, Option<String>>()
```
A similar one but with a function to test type inference by using the ADT constructors would be nice

JorgeCastilloPrz commented on modules/docs/arrow-docs/docs/docs/generic/coproduct/README.md:
````diff
@@ -37,6 +38,18 @@ fun toDisplayValues(items: List<Coproduct3<Car, Dealership, Salesperson>>): List
+}
 ```
 
 Let's say we have an api. Our api operates under the following conditions:
````
I think we could also explain here what's the most effective benefit we've found of Coproducts over sealed classes: to flat the hierarchy. The alternative is composed sealed classes, which doesn't leave you in a pretty neat scenario (sealed classes don't support inheritance outside of their own hierarchy). Describing both scenarios and how they differ would be very enlightening for the reader.

JorgeCastilloPrz commented:
@raulraja don't you think the blogpost style would work good for some ideas here like what coproducts provide versus nested sealed classes? or would you completely move everything to Kdoc?

raulraja commented:
Yes but that also goes well in the class header unless it's tutorial style in which case is fine in the regular docs and should be linked from the class kdoc headers

abergfeld commented:
@JorgeCastilloPrz @raulraja This PR currently has docs being generated with the code:
```
/**
 * Runs the function related to the actual value of the Coproduct and returns the result
 *
 * Parameters: A function for every possible value the Coproduct could be.
 * Returns: RESULT generated by the input functions */
fun <A, B, C, RESULT> Coproduct3<A, B, C>.fold(
    a: (A) -> RESULT,
    b: (B) -> RESULT,
    c: (C) -> RESULT
): RESULT = when (this) {
    is First -> a(this.a)
    is Second -> b(this.b)
    is Third -> c(this.c)
}

```

Is that sufficient to auto link to the API DOCs section?

@raulraja Does the runnable code examples in the docs run using only the latest master version of Arrow? Even compiling the docs locally seems like it should work but they don't run correctly as if they're using the current master version of Coproduct

abergfeld commented:
@JorgeCastilloPrz I think I addressed all your comments except removing the extension functions. I'm not sure if we should just remove those functions now or not.

I've updated the kdocs to be more inlined with what Option is doing and the locally generated docs look like it's all linking up correctly and showing descriptions / params / returns. 

The only part I'm still not sure works is the runnable code examples but it seems like those are just based off the master version of arrow and the examples compile so I'd assume they'll work once merged?

Here's the generated output for Coproduct2 since it's generated and doesn't show as part of the diff:
```kotlin
package arrow.generic.coproduct2

import arrow.core.Option
import arrow.core.toOption
import kotlin.Deprecated
import kotlin.Suppress
import kotlin.Unit

/**
 * Represents a sealed hierarchy of 2 types where only one of the types is actually present.
 */
sealed class Coproduct2<A, B>

/**
 * Represents the first type of a Coproduct2
 */
data class First<A, B>(val a: A) : Coproduct2<A, B>()

/**
 * Represents the second type of a Coproduct2
 */
data class Second<A, B>(val b: B) : Coproduct2<A, B>()

@Deprecated(message = "This has issues with type inference, use First() instead",
replaceWith = ReplaceWith("First<A, B>(a)", "arrow.generic.coproduct2.First")
)
fun <A, B> coproductOf(a: A): Coproduct2<A, B> = First(a)

@Deprecated(message = "This has issues with type inference, use Second() instead",
replaceWith = ReplaceWith("Second<A, B>(b)", "arrow.generic.coproduct2.Second")
)
@Suppress("UNUSED_PARAMETER")
fun <A, B> coproductOf(b: B, dummy0: Unit = Unit): Coproduct2<A, B> = Second(b)

@Deprecated(message = "This has issues with type inference, use .first() instead",
replaceWith = ReplaceWith("this.first<A, B>()")
)
fun <A, B> A.cop(): Coproduct2<A, B> = coproductOf<A, B>(this)

@Deprecated(message = "This has issues with type inference, use .second() instead",
replaceWith = ReplaceWith("this.second<A, B>()")
)
@Suppress("UNUSED_PARAMETER")
fun <A, B> B.cop(dummy0: Unit = Unit): Coproduct2<A, B> = coproductOf<A, B>(this)

/**
 * Creates a Coproduct from the A type
 *
 * @return A Coproduct2<A, B> where the receiver is the A
 */
fun <A, B> A.first(): Coproduct2<A, B> = First(this)

/**
 * Creates a Coproduct from the B type
 *
 * @return A Coproduct2<A, B> where the receiver is the B
 */
fun <A, B> B.second(): Coproduct2<A, B> = Second(this)

/**
 * Transforms the Coproduct into an Option based on the actual value of the Coproduct
 *
 * @return None if the Coproduct was not the specified type, Some if it was the specified type
 */
fun <A> Coproduct2<A, *>.select(): Option<A> = (this as? First)?.a.toOption()

/**
 * Transforms the Coproduct into an Option based on the actual value of the Coproduct
 *
 * @return None if the Coproduct was not the specified type, Some if it was the specified type
 */
@Suppress("UNUSED_PARAMETER")
fun <B> Coproduct2<*, B>.select(dummy0: Unit = Unit): Option<B> = (this as? Second)?.b.toOption()

/**
 * Runs the function related to the actual value of the Coproduct and returns the result
 *
 * @param a The function used to map A to the RESULT type
 * @param b The function used to map B to the RESULT type
 *
 * @return RESULT generated by one of the input functions
 */
fun <A, B, RESULT> Coproduct2<A, B>.fold(a: (A) -> RESULT, b: (B) -> RESULT): RESULT = when (this) {
    is First -> a(this.a)
    is Second -> b(this.b)
}

```

JorgeCastilloPrz commented:
Deprecating the cop() functions makes sense 👍

raulraja commented on modules/docs/arrow-docs/docs/docs/generic/coproduct/README.md:
```diff
@@ -163,10 +235,13 @@ Here we're able to return the result of the `fold` and since it's exhaustively e
 
 We're able to take a Coproduct and `select` the type we care about from it. `select` returns an `Option`, if the value of the Coproduct was for the type you're trying to `select`, you'll get `Some`, if it was not the type used with `select`, you'll get `None`.
 
+{: data-executable='true'}
```
```suggestion
```
This is not needed if we use `kotlin:ank:playground`

raulraja commented on modules/docs/arrow-docs/docs/docs/generic/coproduct/README.md:
````diff
@@ -163,10 +235,13 @@ Here we're able to return the result of the `fold` and since it's exhaustively e
 
+{: data-executable='true'}
+
 ```kotlin:ank
````
```suggestion
```kotlin:ank:playground
```

raulraja commented on modules/docs/arrow-docs/docs/docs/generic/coproduct/README.md:
```diff
@@ -178,16 +253,28 @@ fun handleApiResult(
             )
 }
+
+fun main() {
```
you can use `//sampleStart` and `//sampleEnd` to remove noise from the examples such as imports main etc.

raulraja commented on modules/docs/arrow-docs/docs/docs/generic/coproduct/README.md:
```diff
@@ -178,16 +253,28 @@ fun handleApiResult(
 
 `select` can only be called with a type that exists on the Coproduct, if the type doesn't exist, it won't compile:
+
+{: data-executable='true'}
```
```suggestion

```

raulraja commented on modules/docs/arrow-docs/docs/docs/generic/coproduct/README.md:
````diff
@@ -178,16 +253,28 @@ fun handleApiResult(
+
+{: data-executable='true'}
+
 ```kotlin:ank
````
```suggestion
```kotlin:ank:fail
```
You can use `fail` to let it fail and display the compilation errors in the text.

raulraja commented on modules/docs/arrow-docs/docs/docs/generic/coproduct/README.md:
```diff
@@ -178,16 +253,28 @@ fun handleApiResult(
-//    apiResult.select<String>()
-//error: type mismatch: inferred type is Coproduct3<CommonServerError, RegistrationError, Registration> but Coproduct3<String, *, *> was expected
+fun main() {
+//    println(
```
if you end up using `fail` you'd need to uncomment these all these `//`

raulraja commented on modules/meta/arrow-meta/src/main/java/arrow/generic/CoproductFileGenerator.kt:
```diff
@@ -171,6 +220,28 @@ private fun FileSpec.Builder.addFoldFunction(generics: List<String>): Unit {
+    return "$description\n\n$parameters@return $output\n"
+}
+
+private fun String.toCamelCase(): String {
```
Nothing wrong with this implementation style but soon we will enforce across all arrow modules a style that is more succinct and expression oriented:
```kotlin
private fun String.toCamelCase(): String =
  if (isEmpty()) this 
  else first().toLowerCase() + substring(1)
```
Also this function may be public and placed in the arrow meta String utils file.

abergfeld commented:
@raulraja I think I addressed all your comments 🎉 Let me know if I should update anything else!

JorgeCastilloPrz commented:
So no plans on generating runnable snippets with dokka anymore? Can we just merge like this @raulraja? (coproduct code is compile time generated, we'd get those snippets generated 26 times, which might be too much)

raulraja commented:
it's fine to get the snippets generated N times. All pages need to documented and some users may land straight in coproduct5 for example.

JorgeCastilloPrz commented:
Can we get this in on 0.9.0 or 0.9.1 or something?

raulraja commented:
This is coming out in 0.9.0. @nomisRev and I are making some final touches to Fx and then we will release. Thanks for your patience.

pakoito commented:
@raulraja make sure to double-check the crash!

raulraja commented:
any way to repro on tests or does it require an Android runtime? I may need help with this once fx is ready. I'll ping one of you guys that would not have to download the internet to test it on Android.

JorgeCastilloPrz commented:
What crash are we talking about?
